docs/developers/api-versioning.md

```diff
@@ -0,0 +1,166 @@
+# API Versioning
+
+## Adding feature gated API fields
+
+We've introduced a feature-flag called `enable-api-fields` to the
+[config-feature-flags.yaml file](../../config/config-feature-flags.yaml)
+deployed as part of our releases.
+
+This field can be configured either to be `alpha` or `stable`. This field is
+documented as part of our
+[install docs](../install.md#customizing-the-pipelines-controller-behavior).
+
+For developers adding new features to Pipelines' CRDs we've got a couple of
+helpful tools to make gating those features simpler and to provide a consistent
+testing experience.
+
+### Guarding Features with Feature Gates
+
+Writing new features is made trickier when you need to support both the existing
+stable behaviour as well as your new alpha behaviour.
+
+In reconciler code you can guard your new features with an `if` statement such
+as the following:
+
+```go
+alphaAPIEnabled := config.FromContextOrDefaults(ctx).FeatureFlags.EnableAPIFields == "alpha"
+if alphaAPIEnabled {
+  // new feature code goes here
+} else {
+  // existing stable code goes here
+}
+```
+
+Notice that you'll need a context object to be passed into your function for
+this to work. When writing new features keep in mind that you might need to
+include this in your new function signatures.
+
+### Guarding Validations with Feature Gates
+
+Just because your application code might be correctly observing the feature gate
+flag doesn't mean you're done yet! When a user submits a Tekton resource it's
+validated by Pipelines' webhook. Here too you'll need to ensure your new
+features aren't accidentally accepted when the feature gate suggests they
+shouldn't be. We've got a helper function,
+[`ValidateEnabledAPIFields`](../../pkg/apis/pipeline/v1beta1/version_validation.go),
+to make validating the current feature gate easier. Use it like this:
+
+```go
+requiredVersion := config.AlphaAPIFields
+// errs is an instance of *apis.FieldError, a common type in our validation code
+errs = errs.Also(ValidateEnabledAPIFields(ctx, "your feature name", requiredVersion))
```

Comment:
I know this is not related to this PR, but do we have one for the beta value as well ?

Reply:
this helper can be used for beta API fields as well.

Reply:
it's here now https://github.com/tektoncd/pipeline/blob/main/pkg/apis/version/version_validation.go